python_modules/dagster/dagster/components/core/decl.py

```diff
@@ -274,37 +314,81 @@ def get_component_decl_from_python_file(
-        return ComponentLoaderDecl(context=context, component_node_fn=component_loader)
+        return ComponentLoaderDecl(
+            context=context,
+            component_node_fn=component_loader,
+            path=ComponentPath(file_path=context.path, instance_key=None),
+        )
     else:
         return CompositePythonDecl(
+            path=ComponentPath(file_path=context.path, instance_key=None),
             context=context,
             decls={
-                attr: ComponentLoaderDecl(context=context, component_node_fn=component_loader)
+                attr: ComponentLoaderDecl(
+                    context=context,
+                    component_node_fn=component_loader,
+                    path=ComponentPath(file_path=context.path, instance_key=attr),
+                )
                 for attr, component_loader in component_loaders
             },
         )
 
 
-def load_yaml_component_decl(context: ComponentLoadContext) -> ComponentDecl:
+def build_component_decl_from_yaml_file_backcompat(context: ComponentLoadContext) -> ComponentDecl:
     component_def_path = check.not_none(find_defs_or_component_yaml(context.path))
-    return get_component_decl_from_yaml_file(context=context, component_def_path=component_def_path)
+    return build_component_decl_from_yaml_file(
+        context=context, component_def_path=component_def_path
+    )
 
 
-def get_component_decl_from_yaml_file(
+def build_component_decl_from_yaml_file(
     context: ComponentLoadContext, component_def_path: Path
 ) -> ComponentDecl:
     source_trees = parse_yamls_with_source_position(
         component_def_path.read_text(), str(component_def_path)
     )
     component_nodes = []
-    for source_tree in source_trees:
-        component_nodes.append(YamlDecl.from_source_tree(context=context, source_tree=source_tree))
+    for i, source_tree in enumerate(source_trees):
+        component_nodes.append(
+            build_component_decl_from_yaml_document(
+                context=context,
+                source_tree=source_tree,
+                path=ComponentPath(file_path=context.path, instance_key=i),
+            )
+        )
 
     check.invariant(len(component_nodes) > 0, "No components found in YAML file")
     return YamlFileDecl(
+        path=ComponentPath(file_path=context.path, instance_key=None),
         context=context,
         decls=component_nodes,
         source_positions=[
             source_tree.source_position_tree.position for source_tree in source_trees
         ],
     )
+
+
+def build_component_decl_from_yaml_document(
+    context: ComponentLoadContext,
+    source_tree: ValueAndSourcePositionTree,
+    path: ComponentPath,
+) -> ComponentDecl:
+    component_file_model = _parse_and_populate_model_with_annotated_errors(
+        cls=ComponentFileModel, obj_parse_root=source_tree, obj_key_path_prefix=[]
+    )
+    if component_file_model.type == "dagster.DefsFolderComponent":
```

Comment:
were going to have to do a load on these module strings to see if its a subclass if we don't change things up. Maybe thats fine, were going to end up doing anyway in most cases. 

[1] we should maybe actually do the load and include the loaded type as part of the decl

Reply:
if we dont want to load it we need something to discern in the yaml
* `type: ...` -> `container_type: ...` / `folder_type: ...`
* top level field sibling to `type` that marks as container / folder